net: Prevent node from binding to the same `CService`

w0xlt opened the pull request:
Currently, if the node inadvertently starts with repeated `-bind` options (e.g. `./build/bin/bitcoind -listen -bind=0.0.0.0 -bind=0.0.0.0`), the user will receive a misleading message followed by the node shutdown:

```
[net:error] Unable to bind to 0.0.0.0:8333 on this computer. Bitcoin Core is probably already running.
[error] Unable to bind to 0.0.0.0:8333 on this computer. Bitcoin Core is probably already running.
```

And the user might spend some time looking for a `bitcoind` process or what application is using port 8333, when what happens is that Bitcoin Core successfully connected to port 8333 and then tries again, generating this fatal error.

This PR proposes that repeated `-bind` options have no effect.

jonatack reviewed (commented):
Concept ACK

> This PR proposes that repeated -bind options have no effect.

Only no effect for repeated -bind options having the same value, yes? Would it be preferable to give the user feedback instead?

Would be good to add test coverage.

yuvicc commented:
Concept ACK

l0rinc commented on src/net_permissions.h:
```diff
@@ -84,6 +84,10 @@ class NetWhitebindPermissions : public NetPermissions
     CService m_service;
+
+    bool operator<(const NetWhitebindPermissions& other) const {
+        return m_service < other.m_service;
```
isn't it confusing that in case of service duplicates the it's not obvious which `m_flags` is ignored?

What do we expect to happen in case of something like `-whitebind=127.0.0.1:8335=relay,noban -whitebind=127.0.0.1:8335=relay` vs `-whitebind=127.0.0.1:8335=relay -whitebind=127.0.0.1:8335=relay,noban`?
Or even worse, `-whitebind=127.0.0.1:8335=all -whitebind=127.0.0.1:8335=relay,mempool`?

l0rinc commented on src/net_permissions.h:
```diff
@@ -84,6 +84,10 @@ class NetWhitebindPermissions : public NetPermissions
     static bool TryParse(const std::string& str, NetWhitebindPermissions& output, bilingual_str& error);
     CService m_service;
+
+    bool operator<(const NetWhitebindPermissions& other) const {
```
We could modernize this to C++20
```C++
std::weak_ordering operator<=>(NetWhitebindPermissions const& other) const
```

nit: formatting is off

l0rinc reviewed (changes requested):
Concept ACK, but approach NACK 132a59698e2df2e25fdbae3a6fba308b131419d2

I only did a code review pass on it, please let me know if my concerns are off, happy to ack if I was mistaken.
My biggest concern currently is silently dropping values by partial duplication, e.g. `-whitebind=[::1]:8335=relay -whitebind=[::1]:8335=relay,addr`.

l0rinc commented:
Since these lists are tiny, Instead of a set for deduplication, we might as well use a simple O^2 algorithm, iterate from the back and remove whatever appeared in the front (it will likely be even faster than a set). If you decide to do that, I'd appreciate a randomized unit test checking it against a set deduplication.

l0rinc commented on src/init.cpp:
```diff
@@ -1233,6 +1233,94 @@ bool CheckHostPortOptions(const ArgsManager& args) {
+            [&seen](const CService& service) {
+                return !seen.insert(service).second;  // Remove if already seen
+            });
+        onion_binds.erase(it, onion_binds.end());
```
I find it a bit inconsistent that we're not erring here - we don't know why they provided the same thing twice, seems like a mistake, not sure we should attempt to correct it silently. Would also simplify the code if we unified the way we handle these

l0rinc reviewed (changes requested):
Concept ACK

I personally would not try to correct the behavior (especially silently), given that it's rare, I would simply give the user a warning or error and let them fix the arguments.

I also think the tests are a bit complicated and more stateful now, I have added a few suggestions, maybe we can simplify based on them.

w0xlt committed:
net: Prevent node from binding to the same CService

Currently, if the user inadvertently starts the node with duplicate bind options,
such as `-bind=0.0.0.0 -bind=0.0.0.0`, it will cause a fatal error with the
misleading message "Bitcoin Core is probably already running".

This commit adds early validation to detect duplicate bindings across all binding
configurations (-bind, -whitebind, and onion bindings) before attempting to bind.
When duplicates are detected, the node terminates with a clear, specific error
message: "Duplicate binding configuration for address <addr>. Please check your
-bind, -bind=...=onion and -whitebind settings."

The validation catches duplicates both within the same option type (e.g.,
`-bind=X -bind=X`) and across different types (e.g., `-bind=X -whitebind=Y@X`),
helping users identify and fix configuration mistakes.

w0xlt commented:
@l0rinc thank you for your detailed reviews.

> I personally would not try to correct the behavior (especially silently), given that it's rare, I would simply give the user a warning or error and let them fix the arguments.

I adjusted the approach to meet this requirement. Now when duplicates are detected, the node terminates with a clear, specific error message: "_Duplicate binding configuration for address < addr >. Please check your -bind, -whitebind, and onion binding settings._"

I don't have a strong opinion on whether the node should try to correct the behavior or not. For me, the main issue is the misleading message.

This version is much more simplified compared to the previous ones

l0rinc reviewed (commented):
I like this simpler version more, left a few more comments

Now that we're not deduplicating anymore, the title could be adjusted:
> `net: reject duplicate bind addresses across -bind/-whitebind/onion`

The PR desciption also hints that we still do:
> This PR proposes that repeated -bind options have no effect.

yuvicc commented on src/init.cpp:
```diff
@@ -2142,6 +2176,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
     connOptions.m_i2p_accept_incoming = args.GetBoolArg("-i2pacceptincoming", DEFAULT_I2P_ACCEPT_INCOMING);
 
+    if (auto conflict = CheckBindingConflicts(connOptions)) {
```
I guess the first err msg is from logging while the second is from `init.cpp`. From the comment https://github.com/bitcoin/bitcoin/pull/33231#issuecomment-3222458339, I can see that we would want to stop the execution.

w0xlt replied:
The execution will be stopped either way (in the master branch or in this PR).
This PR simply prevents the node from attempting to connect twice to the same address and port, and instead terminates it with a clear message — rather than the confusing one currently shown in the master branch.

> I guess the first err msg is from logging while the second is from init.cpp

Yes, that makes sense. I'll check again.

l0rinc replied:
it just continues IBD on my side after the warning, it doesn't seem to stop after the error

l0rinc replied:
`feature_bind_extra` is skipped on a Mac

w0xlt replied:
If you try another condition that also triggers `InitError()`, for example:
```
build/bin/bitcoind -i2psam=invalid_addr
```
what happens on your side?

cedwies replied:
Was not yet able to reproduce the continuation. But just to clarify, @l0rinc : are we talking about a late-fail with “afterglow” or does the node truly keep running?

l0rinc replied:
Opened https://github.com/bitcoin/bitcoin/issues/33276, please resolve this comment

naiyoma reviewed (commented):
Concept ACK

Tested with these 3 options:

- `./build/bin/bitcoind -bind=0.0.0.0=onion -bind=0.0.0.0=onion`
- `./build/bin/bitcoind -bind=0.0.0.0 -bind=0.0.0.0 `
- `./build/bin/bitcoind -whitebind=relay@127.0.0.1:8333 -whitebind=mempool@127.0.0.1:8333`


<details>
<summary>bitcoind shuts down  after the error</summary>
[error] Duplicate binding configuration for address 0.0.0.0:8334. Please check your -bind, -whitebind, and onion binding settings.
Error: Duplicate binding configuration for address 0.0.0.0:8334. Please check your -bind, -whitebind, and onion binding settings.
2025-08-28T13:58:57Z torcontrol thread start
2025-08-28T13:58:57Z tor: Thread interrupt
2025-08-28T13:58:57Z Shutdown in progress...
</details>

For duplicate onion binds, there's an existing warning that appears before the error: 
`2025-08-28T13:58:57Z [warning] More than one onion bind address is provided. `
IMO this warning is sufficient, but if you keep the error approach, consider removing the warning.

w0xlt commented:
> IMO this warning is sufficient, but if you keep the error approach, consider removing the warning.
Thanks for the review @naiyoma .
The error will occur regardless, whether in master or in this PR.
This PR simply provides a clearer and more useful message.

Regarding removing the warning, I think the `bind` validation could be moved to occur before the onion validation. Good catch.

vasild commented on src/init.cpp:
```diff
@@ -2142,6 +2176,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     connOptions.m_i2p_accept_incoming = args.GetBoolArg("-i2pacceptincoming", DEFAULT_I2P_ACCEPT_INCOMING);
 
+    if (auto conflict = CheckBindingConflicts(connOptions)) {
+        return InitError(Untranslated(strprintf("Duplicate binding configuration for address %s. "
```
> since we already know which container the duplicate came from

That is known inside `CheckBindingConflicts()` but the info is not signaled to the caller of the function. Also, even inside the function, only one of the containers is known, the other is not. E.g. if the 3rd for-loop finds a duplicate, then it is unknown whether the entry was added in the first or the second for-loop.

> I think this should at least be translatable for consistency with nearby init errors

Right, +1 for translating it.

vasild reviewed (approved):
ACK 28fb7f74c4ae66dcb0f918774d644ac3c393f25b

Some non-blocker comments below, would be happy to re-review if addressed.

w0xlt commented on test/functional/feature_bind_extra.py:
```diff
@@ -87,5 +88,26 @@ def run_test(self):
+        self.stop_node(0)
+
+        # Test all combinations of duplicate bindings (each should fail)
+        error_msg = "Error: Duplicate binding configuration for address 127.0.0.1:11012. " \
```
Done in 1c40b32597712059d5d809925da0e9adccac0fb3. Thanks.

w0xlt commented on test/functional/feature_bind_extra.py:
```diff
@@ -87,5 +88,26 @@ def run_test(self):
+                    self.nodes[0].assert_start_raises_init_error(
+                        [opt1, opt2],
+                        error_msg,
+                        match=ErrorMatch.FULL_TEXT)
```
Done in 1c40b32597712059d5d809925da0e9adccac0fb3.

w0xlt commented on test/functional/feature_bind_extra.py:
```diff
@@ -11,6 +11,7 @@
     addr_to_hex,
     get_bind_addrs,
 )
+from test_framework.test_node import ErrorMatch
```
Done in 1c40b32597712059d5d809925da0e9adccac0fb3.

w0xlt commented on src/init.cpp:
```diff
@@ -1233,6 +1233,40 @@ bool CheckHostPortOptions(const ArgsManager& args) {
+ *
+ * @param connOptions Connection options containing the binding vectors to check
+ * @return std::optional<CService> containing the first duplicate found, or std::nullopt if no duplicates
+ */
```
I’ll leave the comment in for now. 
I agree that clear, organized code is the best form of documentation, but I think helpful comments can still save valuable developer time.

w0xlt commented on src/init.cpp:
```diff
@@ -1233,6 +1233,40 @@ bool CheckHostPortOptions(const ArgsManager& args) {
+/**
+ * @brief Checks for duplicate bindings across all binding configuration.
+ *
+ * @param connOptions Connection options containing the binding vectors to check
```
Done in https://github.com/bitcoin/bitcoin/commit/1c40b32597712059d5d809925da0e9adccac0fb3.

w0xlt commented on src/init.cpp:
```diff
@@ -1233,6 +1233,40 @@ bool CheckHostPortOptions(const ArgsManager& args) {
+ * @param connOptions Connection options containing the binding vectors to check
+ * @return std::optional<CService> containing the first duplicate found, or std::nullopt if no duplicates
+ */
+std::optional<CService> CheckBindingConflicts(const CConnman::Options& connOptions)
```
Done in https://github.com/bitcoin/bitcoin/commit/1c40b32597712059d5d809925da0e9adccac0fb3.

w0xlt commented on src/init.cpp:
```diff
@@ -2142,6 +2176,12 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
+    if (auto conflict = CheckBindingConflicts(connOptions)) {
+        return InitError(Untranslated(strprintf("Duplicate binding configuration for address %s. "
+            "Please check your -bind, -whitebind, and onion binding settings.",
```
Done in https://github.com/bitcoin/bitcoin/commit/1c40b32597712059d5d809925da0e9adccac0fb3. Thanks for the review.

yuvicc reviewed (commented):
ACK 1c40b32597712059d5d809925da0e9adccac0fb3

`bitcoind` stops with error message for duplicate bind options. 
```
Error: Duplicate binding configuration for address 0.0.0.0:8333. Please check your -bind, -bind=...=onion and -whitebind settings.
2025-09-05T05:15:05Z torcontrol thread start
2025-09-05T05:15:05Z Loading 0 mempool transactions from file...
2025-09-05T05:15:05Z Imported mempool transactions from file: 0 succeeded, 0 failed, 0 expired, 0 already there, 0 waiting for initial broadcast
2025-09-05T05:15:05Z initload thread exit
```

l0rinc commented on test/functional/feature_bind_extra.py:
```diff
@@ -87,5 +88,11 @@ def run_test(self):
+
+        addr = "127.0.0.1:11012"
+        for opt1, opt2 in combinations_with_replacement([f"-bind={addr}", f"-bind={addr}=onion", f"-whitebind=noban@{addr}"], 2):
+            print(f"opt1 = {opt1}, opt2 = {opt2}") # replace with assert, I just used this to validate if the iteration is the same
```
@yuvicc, this still needs fixing

l0rinc commented:
ACK 4d4789dffad55b96f1cb96b718cc6923f5344454

vasild reviewed (approved):
ACK 4d4789dffad55b96f1cb96b718cc6923f5344454

naiyoma reviewed (commented):
Tested ACK 4d4789dffad55b96f1cb96b718cc6923f5344454

Manually tested with different duplicate bindings.
```
./build/bin/bitcoind -bind=0.0.0.0=onion -bind=0.0.0.0=onion
./build/bin/bitcoind -bind=0.0.0.0 -bind=0.0.0.0
./build/bin/bitcoind -whitebind=relay@127.0.0.1:8333 -whitebind=mempool@127.0.0.1:8333
```
On master, the error message is misleading
```
2025-09-09T10:59:38Z [net:error] Unable to bind to 0.0.0.0:8334 on this computer. Bitcoin Core is probably already running.
2025-09-09T10:59:38Z [error] Unable to bind to 0.0.0.0:8334 on this computer. Bitcoin Core is probably already running.
Error: Unable to bind to 0.0.0.0:8334 on this computer. Bitcoin Core is probably already running.
```

After this PR change, the error message is much clearer and points to the issue.

```
2025-09-09T11:02:19Z [error] Duplicate binding configuration for address 0.0.0.0:8334. Please check your -bind, -bind=...=onion and -whitebind settings.
Error: Duplicate binding configuration for address 0.0.0.0:8334. Please check your -bind, -bind=...=onion and -whitebind settings.
```

achow101 commented:
ACK 4d4789dffad55b96f1cb96b718cc6923f5344454
